### PR TITLE
Implements request body validation inside HmacAuth

### PR DESCRIPTION
**Pull request by vaibhavatul47**
### Summary
Adding functionality to validate request body in hmac-auth. This is optional and can be configured when enabling the Hmac-Auth plugin for an API.

### Full changelog

Currently, Hmac-Auth plugin allows validation of only request headers. There are requirements in real life projects when the validity of request body is also to be verified. To solve this problem I have introduced the usage of digest header. This digest header will contain the Base64 encoded Sha256 hash of the request body.

If an API has hmac-auth plugin enabled and validate-request-body config is enabled then for every request of this API received at KONG, Hmac-Auth plugin will also validate the integrity of request body, needless to say, this check will be performed only if request contains any body

If this config( validate-request-body) is enabled on an API then every request must contain digest header inside hmac-headers parameter.

### Tests added

* should pass with GET when body validation enabled
* should pass with POST when body validation is enabled and digest header present
* should not pass with POST when body validation enabled and digest header missing
* ~~should not pass with POST when body validation enabled and digest header missing from hmac-headers~~ ( This test will be added in #2418 )
* should not pass with POST when body validation enabled and postBody is tampered
* should not pass with POST when body validation enabled and digest header is tampered

### CODE REVIEW

**Comment by vaibhavatul47** [comment]
This will fail in some tests as those depends on changes in PR: https://github.com/Mashape/kong/pull/2418
So those tests will pass once PR #2418 will be merged into this

**Comment by p0pr0ck5** [comment]
@vaibhavatul47 please address the linting error and test failures: https://travis-ci.org/Mashape/kong/builds/223708709

**Comment by p0pr0ck5** [comment]
@vaibhavatul47 there is still a failure in the test:

` Plugin: hmac-auth (access) HMAC Authentication should not pass with POST when body validation enabled and digest header missing from hmac-headers`

Mind taking a look?

**Comment by vaibhavatul47** [comment]
@p0pr0ck5 This is expected behaviour as of now. Please also have a look at [this comment](https://github.com/Mashape/kong/pull/2418#issuecomment-298495045) on #2418

**Comment by vaibhavatul47** [comment]
Hi @p0pr0ck5, I was busy in personal work so couldn't finish this sooner.
I have made required changes for this PR, i.e. I have removed failed tests from this PR. I would add them in #2418 once this PR is merged.

**Comment by thibaultcha** [comment]
@p0pr0ck5 and @shashiranjan84 will review this, as it is in need of a few improvements :) We will also need a documentation PR on getkong.org :) Thanks!

File: kong/plugins/hmac-auth/access.lua [hunk 1]
```diff
@@ -82,11 +87,26 @@ local function create_hash(request, hmac_params, headers)
   return ngx_sha1(hmac_params.secret, signing_string)
 end
 
+local function is_signature_equal(signature_1, signature_2)
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: what is the purpose of this function? constant time comparison? if so, please see 7345134740dbb01057b9d69f1f8bc91019a20378 as to why this approach is incorrect; otherwise, please elaborate :)

File: kong/plugins/hmac-auth/access.lua [hunk 2]
```diff
@@ -152,7 +188,7 @@ local function set_consumer(consumer, credential)
   else
     ngx_set_header(constants.HEADERS.ANONYMOUS, true)
   end
-  
+
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: this change seems unwarranted? please remove it.

File: kong/plugins/hmac-auth/access.lua [hunk 3]
```diff
@@ -173,10 +209,21 @@ local function do_authentication(conf)
   if not hmac_params.username then
     hmac_params = retrieve_hmac_fields(ngx.req, headers, AUTHORIZATION, conf)
   end
+
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: idem

File: kong/plugins/hmac-auth/access.lua [hunk 4]
```diff
@@ -189,7 +236,7 @@ local function do_authentication(conf)
 
   -- Retrieve consumer
   local consumer, err = cache.get_or_set(cache.consumer_key(credential.consumer_id),
-                   nil, load_consumer_into_memory, credential.consumer_id)
+    nil, load_consumer_into_memory, credential.consumer_id)
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: this whitespace change is incorrect, please remove it.

File: kong/plugins/hmac-auth/access.lua [hunk 5]
```diff
@@ -199,11 +246,10 @@ local function do_authentication(conf)
   return true
 end
 
-
 function _M.execute(conf)
 
   if ngx.ctx.authenticated_credential and conf.anonymous ~= "" then
-    -- we're already authenticated, and we're configured for using anonymous, 
+    -- we're already authenticated, and we're configured for using anonymous,
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: please avoid whitespace changes like this, as it dirties the commit history and makes future archaeological efforts more challenging.

File: kong/plugins/hmac-auth/access.lua [hunk 6]
```diff
@@ -213,7 +259,7 @@ function _M.execute(conf)
     if conf.anonymous ~= "" and conf.anonymous ~= nil then
       -- get anonymous user
       local consumer, err = cache.get_or_set(cache.consumer_key(conf.anonymous),
-                       nil, load_consumer_into_memory, conf.anonymous, true)
+        nil, load_consumer_into_memory, conf.anonymous, true)
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: idem, please remove this unnecessary whitespace change.

File: kong/plugins/hmac-auth/access.lua [hunk 7]
```diff
@@ -129,6 +149,22 @@ local function validate_clock_skew(headers, date_header_name, allowed_clock_skew
   return true
 end
 
+local function validate_request_body(headers, body)
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: The ietf draft indicates that a digest header may look something like:

`digest: SHA-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=\n`

It doesn't appear that the generate hash will match such a format, with the form of the hash prepended to the base64-encoded hash itself. Do we care about this case?

Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: I wrote my logic assuming Kong will always check the SHA-256 hash of request body.
As per ieft draft, the user can use any hash algorithm to generate the hash. If we allow this, then for each request, Kong would need to parse `Digest` header and then figure out which hashing algorithm to use. Then there could be cases where a user uses an algorithm not supported by Kong libraries.
So for simplicity, I assumed only SHA-256.

Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: I can do one thing to abide by ieft draft.
At Kong, we can generate hash string prepended by the hash algorithm as recommended by draft, i.e.: 
`digest: SHA-256=X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=\n`
Then we would need to mention it clearly that user needs to send digest header in this format and only `sha-256` algorithm is supported by Kong.

Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: @vaibhavatul47 please also rebase to `next`.

File: kong/plugins/hmac-auth/access.lua (hunk 1 above)
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: Sorry, I mentioned the wrong commit. I should have typed 6ccdf7b664957edef3f557c53ae706456c411e0d :)

File: kong/plugins/hmac-auth/access.lua (hunk 7 above)
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: @vaibhavatul47 i think this is a good approach, as it could allow us to support additional algorithms in the future.

Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: I've updated this change.

File: kong/plugins/hmac-auth/access.lua (hunk 1 above)
Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: I've removed this function as no more required.

**Comment by p0pr0ck5** [comment]
LGTM, @vaibhavatul47 would you be able to adjust your commit message based on https://github.com/Mashape/kong/blob/master/CONTRIBUTING.md#commit-message-format? @shashiranjan84 any thoughts here?

File: kong/plugins/hmac-auth/access.lua [hunk 8]
```diff
@@ -129,6 +134,22 @@ local function validate_clock_skew(headers, date_header_name, allowed_clock_skew
...
+  local sha_recieved = headers[DIGEST]
+
+  if body == nil then
+    return true
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: style, needs a blank line after return. not a blocker, we can adjust this as part of merge

**Comment by shashiranjan84** [comment]
@vaibhavatul47 @p0pr0ck5 overall LGTM, I'll format it to latest code style before merge.

**Comment by vaibhavatul47** [comment]
@p0pr0ck5 I've updated commit message.

File: kong/plugins/hmac-auth/access.lua [hunk 9]
```diff
@@ -129,6 +134,22 @@ local function validate_clock_skew(headers, date_header_name, allowed_clock_skew
...
+  elseif sha_recieved == nil then
+    return false
```
Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: @vaibhavatul47 here your logic forcing client to send `digest` header. I would prefer logic to be 
```
if validate_request_body and headers[digest] then
  validate digest
end
```

Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: I ll update the code and tests

Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: @vaibhavatul47  here is what I am thinking to change it to 
```
-- If request body validation is enabled, request may require `digest` validation 
 local digest_recieved = headers[DIGEST]
  if conf.validate_request_body and digest_recieved then

    if not validate_digest(digest_recieved) then
      return false, {status = 403, message = SIGNATURE_NOT_VALID }
    end
  end


-----------------------
local function validate_digest(headers)

  local sha_recieved = headers[DIGEST]
  -- Client didn't send `digest` header
  -- as it is not interested in body validation
  if not sha_recieved then
    return true
  end

  req_read_body()
  local body = req_get_body_data()
  -- request must have body as client sent
  -- a digest header
  if not body then
    return false
  end

  local sha256 = resty_sha256:new()
  sha256:update(body)
  local sha_created = "SHA-256=" .. ngx_encode_base64(sha256:final())

  return sha_created == sha_recieved
end
```

I will also work on adding support for other hashing algorithm, at least SHA1 and SHA-256 for both Digest and HTTP signature validation

File: kong/plugins/hmac-auth/access.lua [hunk 10]
```diff
@@ -177,6 +198,16 @@ local function do_authentication(conf)
     return false, {status = 403, message = SIGNATURE_NOT_VALID}
   end
 
+  -- If request body validation is enabled, then verify hash of request body.
+  if conf.validate_request_body then
```
Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: Also we should validate headers first and then body. As if header is tampered there is no point validating body.

**Comment by vaibhavatul47** [comment]
@shashiranjan84 if validate_request_body is enabled then forcing client to send `digest` header is important. If this is not mandated then someone who can change request body would also delete digest header and then this corrupted request would be validated by plugin as it didn't contain any digest header.

File: kong/plugins/hmac-auth/access.lua (hunk 9 above)
Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: @shashiranjan84 if validate_request_body is enabled then forcing client to send digest header is important. If this is not mandated then someone who can change request body would also delete digest header and then this corrupted request would be validated by plugin as it didn't contain any digest header.

File: kong/plugins/hmac-auth/access.lua (hunk 10 above)
Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: @shashiranjan84 I'll update code according to this.

File: kong/plugins/hmac-auth/access.lua (hunk 9 above)
Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: @vaibhavatul47 Not really, If signature is created  using digest, it would not pass validation as it is tampered.

File: kong/plugins/hmac-auth/access.lua (hunk 10 above)
Comment (shashiranjan84) [inline kong/plugins/hmac-auth/access.lua]: Please don't, I already in updated the code. Planning to merge it today

Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: ok

File: kong/plugins/hmac-auth/access.lua (hunk 9 above)
Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: 👍

**Comment by shashiranjan84** [comment]
Closing this PR in favor of #2613.

**Comment by p0pr0ck5** [comment]
#2613 was merged, thank you @vaibhavatul47 for your contribution!
